src/client/lazy-app/Compress/Modal/index.tsx

```diff
@@ -0,0 +1,148 @@
+import { h, Component } from 'preact';
+import * as style from './style.css';
+import 'add-css:./style.css';
+import { linkRef } from 'shared/prerendered-app/util';
+import {
+  InfoIcon,
+  DiamondStarIcon,
+  ModalErrorIcon,
+} from 'client/lazy-app/icons';
+
+interface Props {}
+
+export interface ModalMessage {
+  type: 'info' | 'error' | 'update';
+  title: string;
+  content: string;
+}
+
+interface State {
+  message: ModalMessage;
+  shown: boolean;
+}
+
+export default class Modal extends Component<Props, State> {
+  state: State = {
+    message: {
+      type: 'info',
+      title: 'default',
+      content: 'default',
+    },
+    shown: false,
+  };
+
+  private modal?: HTMLElement;
+  private returnFocusElement?: HTMLElement | null;
+
+  showModal(message: ModalMessage) {
+    if (this.state.shown) return;
+    if (!this.modal) return;
+
+    // Set element to return focus to after hiding
+    this.returnFocusElement = document.activeElement as HTMLElement;
+
+    this.modal.style.display = '';
+    this.setState({
+      message: message,
+      shown: true,
+    });
+    // Wait for the 'display' reset to take place, then focus
+    setTimeout(() => {
+      this.modal?.querySelector('button')?.focus();
+    }, 0);
+  }
+
+  hideModal() {
+    this.setState({
+      message: { ...this.state.message },
+      shown: false,
+    });
+    setTimeout(() => {
+      this.modal && (this.modal.style.display = 'none');
+      this.returnFocusElement?.focus();
+    }, 250);
+  }
+
+  private _getCloseButton() {
+    return this.modal!.querySelector('button')!;
+  }
+
+  private _getLastFocusable() {
+    const focusables = this.modal!.querySelectorAll('button, a');
+    return focusables[focusables.length - 1] as HTMLElement;
+  }
+
+  private _onKeyDown(e: KeyboardEvent) {
+    // If Escape, hide modal
+    if (e.key === 'Escape' || e.keyCode == 27) {
+      this.hideModal();
+      e.preventDefault();
+      e.stopImmediatePropagation();
+      return;
+    }
+
+    let isTabPressed = e.key === 'Tab' || e.keyCode === 9;
+
+    if (!isTabPressed) return;
+
+    if (e.shiftKey) {
+      // If SHIFT + TAB was pressed on the first focusable element
+      // Move focus to the last focusable element
+      if (document.activeElement === this._getCloseButton()) {
+        this._getLastFocusable().focus();
+        e.preventDefault();
+        e.stopImmediatePropagation();
+      }
+    } else {
+      // If TAB was pressed on the last focusable element
+      // Move focus to the first focusable element
+      if (document.activeElement === this._getLastFocusable()) {
+        this._getCloseButton().focus();
+        e.preventDefault();
+        e.stopImmediatePropagation();
+      }
+    }
+  }
```

Comment:
On the one hand, I am impressed that you managed to successfully trap focus manually :D

However, I think there are some platform features that we can use to achieve the same thing with less code. The first options is the `inert` attribute, that can make all other HTML in the tree unfocusable. Read [more here](https://developer.chrome.com/articles/inert/). The other option, and the even more natural fit I think, is using the `<dialog>` element, which is now [well supported](https://caniuse.com/dialog).

Reply:
I **did** want to use the `<dialog>` element, but isn't it fairly new on iOS? I personally know lots of people using iOS 14 and who stubbornly refuse to update. But yeah, most user devices should be able to support it, and I'm all for using a cool new web feature. Lemme know what you think!

Reply:
I was iffy about using `inert` because I'd have to apply it to all element siblings of the Modal components. Wouldn't that initiate a huge reflow?